tests/ee/common/tupleschema_test.cpp

```
@@ -0,0 +1,235 @@
+    }
+
+    ~ScopedSchema() {
+        TupleSchema::freeTupleSchema(m_schema);
+    }
+
+private:
+    TupleSchema* m_schema;
+};
+
+TEST_F(TupleSchemaTest, Basic)
+{
+    voltdb::TupleSchemaBuilder builder(2);
+
+    builder.setColumnAtIndex(0, VALUE_TYPE_INTEGER);
+    builder.setColumnAtIndex(1, VALUE_TYPE_VARCHAR,
+                             256, // column size
+                             false, // do not allow nulls
+                             true); // size is in bytes
+
+    ScopedSchema schema(builder.build());
+
+    ASSERT_NE(NULL, schema.get());
+    ASSERT_EQ(2, schema->columnCount());
+
+    EXPECT_EQ(1, schema->getUninlinedObjectColumnCount());
+    EXPECT_EQ(1, schema->getUninlinedObjectColumnInfoIndex(0));
+
+    // 4 bytes for the integer
+    // 8 bytes for the string pointer
+    EXPECT_EQ(12, schema->tupleLength());
+
+    const TupleSchema::ColumnInfo *colInfo = schema->getColumnInfo(0);
+    ASSERT_NE(NULL, colInfo);
+    EXPECT_EQ(0, colInfo->offset);
+    EXPECT_EQ(4, colInfo->length);
+    EXPECT_EQ(VALUE_TYPE_INTEGER, colInfo->type);
+    EXPECT_EQ(1, colInfo->allowNull);
+    EXPECT_EQ(true, colInfo->inlined);
+    EXPECT_EQ(false, colInfo->inBytes);
+
+    colInfo = schema->getColumnInfo(1);
+    ASSERT_NE(NULL, colInfo);
+    EXPECT_EQ(4, colInfo->offset);
+    EXPECT_EQ(256, colInfo->length);
+    EXPECT_EQ(VALUE_TYPE_VARCHAR, colInfo->type);
+    EXPECT_EQ(false, colInfo->allowNull);
+    EXPECT_EQ(false, colInfo->inlined);
+    EXPECT_EQ(true, colInfo->inBytes);
+}
+
+TEST_F(TupleSchemaTest, HiddenColumn)
+{
+    voltdb::TupleSchemaBuilder builder(2,  // 2 visible columns
+                                       1); // 1 hidden column
+    builder.setColumnAtIndex(0, VALUE_TYPE_INTEGER);
+    builder.setColumnAtIndex(1, VALUE_TYPE_VARCHAR,
+                             256,   // column size
+                             false, // do not allow nulls
+                             true); // size is in bytes
```

I see your point.  It would be really confusing to return a constraint violation to the user if a NULL appeared in a hidden column where it shouldn't.  We would want to return an internal error if this ever happened.

However, I think that the existing ColumnInfo structure does an okay job for both visible and hidden columns, and for hidden columns we need to put something in the `allowNulls` field.  By allowing clients of TupleSchema to specify a value here, readers of the code will learn something about how the hidden column is used.  For example, the hidden DR timestamp column should never be null if it exists.